lib/RestructureCFG/SelectScopePass.cpp

```diff
@@ -0,0 +1,261 @@
+    revng_log(SelectScopePassLogger,
+              "  Predecessor: " + Predecessor->getName().str());
+    Predecessors.insert(Predecessor);
+  }
+
+  return Predecessors;
+}
+
+static SmallVector<BasicBlock *> getNodesToProcess(BasicBlock *N,
+                                                   BasicBlock *PostDominator) {
+  // We exploit the `Visited` set, by passing it to
+  // `ReversePostOrderTraversalExt`, in order to stop the visit at the
+  // `PostDominator`
+  std::set<BasicBlock *> Visited;
+  Visited.insert(PostDominator);
+
+  // We collect all the nodes between the conditional `PONode` and its
+  // immediate postdominator, by using the `ReversePostOrderTraversalExt`
+  SmallVector<BasicBlock *> NodesToProcess;
+  for (BasicBlock *RPONode :
+       ReversePostOrderTraversalExt<Scope<BasicBlock *>>(N, Visited)) {
+    NodesToProcess.push_back(RPONode);
+  }
+
+  // From the collected nodes, we need to remove the first node, which
+  // corresponds to the `PONode`, which should not be processed in this
+  // round
+  revng_assert(NodesToProcess.front() == N);
+  NodesToProcess.erase(NodesToProcess.begin());
+
+  return NodesToProcess;
+}
+
+class SelectScopePassImpl {
+  Function &F;
+  PostDomTreeOnView<BasicBlock, Scope> PDT;
+
+  // We keep a boolean field to track whether the `Function` was modified
+  bool FunctionModified = false;
+
+public:
+  SelectScopePassImpl(Function &F) : F(F) {}
+
+public:
+  void processConditionalNode(BasicBlock *ConditionalNode) {
+    SetVector<BasicBlock *>
+      ConditionalSuccessors = getConditionalSuccessors(ConditionalNode);
+
+    // We skip all the nodes which are not conditional
+    if (ConditionalSuccessors.size() <= 1) {
+      return;
+    }
+
+    revng_log(SelectScopePassLogger,
+              "Processing conditional " << ConditionalNode->getName().str()
+                                        << "\n");
+
+    BasicBlock *PostDominator = PDT[ConditionalNode]->getIDom()->getBlock();
+    revng_assert(PostDominator);
+    revng_log(SelectScopePassLogger,
+              "The identified postdominator is "
+                << PostDominator->getName().str() << "\n");
+
+    SmallVector<BasicBlock *>
+      NodesToProcess = getNodesToProcess(ConditionalNode, PostDominator);
+    revng_log(SelectScopePassLogger,
+              "Nodes between conditional and its postdominator, in reverse "
+              "post order:\n");
+    for (auto DFSNode : NodesToProcess) {
+      revng_log(SelectScopePassLogger, "  " << DFSNode->getName().str());
+    }
+
+    // Initialize the `ReachabilityMap`
+    std::map<BasicBlock *, size_t> ReachabilityMap;
+
+    // We initialize the `ReachabilityMap` for each `ConditionalSuccessor`
+    for (const auto &[Index, ConditionalSuccessor] :
+         enumerate(ConditionalSuccessors)) {
+      if (not ReachabilityMap.contains(ConditionalSuccessor)) {
+        ReachabilityMap[ConditionalSuccessor] = Index;
+      }
+    }
```

Comment:
This can go in a separate `static` free function.

Also, it seems to me that the `ReachabilityMap` isn't really a reachability map. It's more of a `ScopeMap`: it maps each key `BasicBlock *` to a scope ID.  And the goal of the bulk of the algorithm below is to assign each node in `NodesToProcess` to one and only one scope ID.